esp8266/modesp.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * This file is part of the Micro Python project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015 Paul Sokolovsky
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdbool.h>
+
+#include "py/nlr.h"
+#include "py/obj.h"
+#include "py/gc.h"
+#include "py/runtime.h"
+#include MICROPY_HAL_H
+#include "queue.h"
+#include "user_interface.h"
+
+STATIC mp_obj_t scan_cb_obj;
+
+STATIC void esp_scan_cb(scaninfo *si, STATUS status) {
+    //printf("in pyb_scan_cb: %d, si=%p, si->pbss=%p\n", status, si, si->pbss);
+    struct bss_info *bs;
+    if (si->pbss) {
+        STAILQ_FOREACH(bs, si->pbss, next) {
+            mp_obj_tuple_t *t = mp_obj_new_tuple(6, NULL);
+            t->items[0] = mp_obj_new_bytes(bs->ssid, strlen((char*)bs->ssid));
+            t->items[1] = mp_obj_new_bytes(bs->bssid, sizeof(bs->bssid));
+            t->items[2] = MP_OBJ_NEW_SMALL_INT(bs->channel);
+            t->items[3] = MP_OBJ_NEW_SMALL_INT(bs->rssi);
+            t->items[4] = MP_OBJ_NEW_SMALL_INT(bs->authmode);
+            t->items[5] = MP_OBJ_NEW_SMALL_INT(bs->is_hidden);
+            mp_call_function_1(scan_cb_obj, t);
```

Comment:
Nope, need to be done. But as I mentioned in https://github.com/micropython/micropython/pull/1068#discussion_r23111702 , it looks that we need to repeat code for callback protection again and again...